pkg/knowledge_base2/emitter.go

```diff
@@ -109,9 +185,16 @@ func (c *ConsumptionObject) Convert(value any, res construct.ResourceId, ctx Dyn
 		return value, err
 	}
 	bstr := strings.TrimSpace(buf.String())
+	// We convert here just to make sure it gets translated to the right type of input
+	// We will convert again when consuming to ensure strings/etc are converted to their respective struct
+	// if they match a property ref/id/etc
 	val, err := TransformToPropertyValue(res, c.PropertyPath, bstr, ctx, DynamicValueData{Resource: res})
-	if err == nil {
-		return val, nil
+	if err != nil {
+		return val, err
+	}
+	val, err = TransformToPropertyValue(res, c.PropertyPath, val, ctx, DynamicValueData{Resource: res})
```

Comment:
The double convert is very weird, why doesn't `TransformtToPropertyValue` do this? I think it would violate the contract of what that function should do if it doesn't return a fully valid, transformed value.

What's the use-case for why this won't work in one shot with the current implementation?

Reply:
when we pass it a json string, it doesnt work in one shot. Thats what im running into now, because it converts the json string to map(string,string) but it doesnt evaluate the value to see if its a property ref or resource id. We can likely fix this function

Reply:
Yeah, looks like it isn't doing the right thing in `MapPropertyType.Parse`. After decoding the string, it immediately returns instead of checking the sub properties.